widget/text.go

```diff
@@ -16,212 +16,365 @@ const (
+func (t *RichText) CreateRenderer() fyne.WidgetRenderer {
+	t.ExtendBaseWidget(t)
+	r := &textRenderer{obj: t}
 
-	t.propertyLock.Lock()
 	t.updateRowBounds() // set up the initial text layout etc
-	t.propertyLock.Unlock()
 	r.Refresh()
 	return r
 }
 
-func (t *textProvider) Resize(size fyne.Size) {
+// MinSize calculates the minimum size of this rich text.
+// This is based on the contained text with a standard amount of padding added.
+func (t *RichText) MinSize() fyne.Size {
+	charMinSize := t.charMinSize(false)
+	concealedMinSize := t.charMinSize(true)
+	height := float32(0)
+	width := float32(0)
+	i := 0
+
+	t.propertyLock.RLock()
+	count := t.rows()
+	wrap := t.Wrapping
+	t.propertyLock.RUnlock()
+
+	for ; i < count; i++ {
+		str := string(t.row(i))
+		bound := t.rowBoundary(i)
+		min := fyne.MeasureText(str, bound.seg.size(), bound.seg.Style.TextStyle)
+		if str == "" {
+			if bound.seg.concealed {
+				min = concealedMinSize
+			} else {
+				min = charMinSize
+			}
+		}
+		if wrap == fyne.TextWrapOff {
+			width = fyne.Max(width, min.Width)
+		}
+		if i == count-1 || bound == nil || !bound.inline {
+			height += min.Height
+		}
+	}
+
+	if height == 0 {
+		height = charMinSize.Height
+	}
+
+	return fyne.NewSize(width, height).
+		Add(fyne.NewSize(theme.Padding()*4, theme.Padding()*4).Subtract(t.inset).Subtract(t.inset))
+}
+
+// Resize sets a new size for the rich text.
+// This should only be called if it is not in a container with a layout manager.
+//
+// Implements: fyne.Widget
+func (t *RichText) Resize(size fyne.Size) {
 	t.propertyLock.RLock()
 	baseSize := t.size
-	presenter := t.presenter
 	t.propertyLock.RUnlock()
 	if baseSize == size {
 		return
 	}
 
 	t.propertyLock.Lock()
 	t.size = size
-
-	t.updateRowBounds()
 	t.propertyLock.Unlock()
+	t.updateRowBounds()
 
-	if presenter != nil {
-		t.refreshTextRenderer()
-		cache.Renderer(t).Layout(size)
-	}
-}
-
-// updateRowBounds updates the row bounds used to render properly the text widget.
-// updateRowBounds should be invoked every time t.buffer or viewport changes.
-func (t *textProvider) updateRowBounds() {
-	if t.presenter == nil {
-		t.rowBounds = [][2]int{}
-		return // not yet shown
-	}
-	textWrap := t.presenter.textWrap()
-	textStyle := t.presenter.textStyle()
-	textSize := theme.TextSize()
-	maxWidth := t.size.Width - 2*theme.Padding()
-
-	t.rowBounds = lineBounds(t.buffer, textWrap, maxWidth, func(text []rune) float32 {
-		return fyne.MeasureText(string(text), textSize, textStyle).Width
-	})
-}
-
-// refreshTextRenderer refresh the textRenderer canvas objects
-// this method should be invoked every time the t.buffer changes
-// example:
-// t.buffer = []rune("new text")
-// t.updateRowBounds()
-// t.refreshTextRenderer()
-func (t *textProvider) refreshTextRenderer() {
-	if t.presenter == nil {
-		return // not yet shown
-	}
-	obj := t.presenter.object()
-	if obj == nil {
-		obj = t
-	}
-
-	obj.Refresh()
+	t.Refresh()
+	cache.Renderer(t).Layout(size)
```

Comment:
this will create a phantom renderer that is not drawn in the canvas. this should be
```
cache.Renderer(t.super()).Layout(size)
```

Or maybe this could be removed, because this does not do anything before (because it created a new renderer for textProvider instead of using the one for the presenter).

Reply:
Good catch thanks, fixed